drivers/at86rf215/at86rf215_o-qpsk.c

```diff
@@ -148,20 +150,75 @@ static inline uint8_t _AGCC(uint8_t chips)
     }
 }
 
-/* Table 6-100. MR-O-QPSK Modes */
-static uint32_t _get_bitrate(uint8_t chips, uint8_t mode)
+static inline uint16_t _get_symbol_duration_us(uint8_t chips)
 {
-    switch (chips) {
-    case BB_FCHIP100:
-        return  6250 * (1 << mode);
-    case BB_FCHIP200:
-        return 12500 * (1 << mode);
-    case BB_FCHIP1000:
-    case BB_FCHIP2000:
-        return mode ? 125000 * (1 << (mode - 1)) : 31250;
+    const uint16_t oqpsk_duration[] = { 320, 160, 64, 64 };
+    return oqpsk_duration[chips];
+}
+
+static inline uint8_t _get_cca_duration_syms(uint8_t chips)
+{
+    return (chips < BB_FCHIP1000) ? 4 : 8;
+}
+
+static inline uint8_t _get_shr_duration_syms(uint8_t chips)
+{
+    return (chips < BB_FCHIP1000) ? 48 : 72;
+}
+
+static uint8_t _get_spreading(uint8_t chips, uint8_t mode)
+{
+    if (mode == 4) {
+        return 1;
     }
 
-    return 0;
+    uint8_t spread = 1 << (3 - mode);
+
+    if (chips == BB_FCHIP1000) {
+        return 2 * spread;
+    }
+
+    if (chips == BB_FCHIP2000) {
+        return 4 * spread;
+    }
+
+    return spread;
+}
+
+static inline uint8_t _get_ack_psdu_duration_syms(uint8_t chips, uint8_t mode)
+{
+    static const uint8_t ack_syms[] = { 32, 32, 64, 128 };
+    const uint8_t syms = ack_syms[chips];
+    const uint8_t spread = _get_spreading(chips, mode);
+    const uint16_t Npsdu = spread * 2 * 63;
+
+    uint16_t duration_sym = (Npsdu + syms/2) / syms;
+    duration_sym += (Npsdu + 8 * syms) / (16 * syms);
```

Comment:
Can you help me understand where all these numbers come from?

Reply:
The dreaded section 18.3.2.14 of 802.15.4g-2012
The Atmel reference driver helped to de-obfuscate the standard a bit.